### PR TITLE
Fixes tablet location cache bug that caused batch scanner to return duplicate data

### PR DESCRIPTION
**Pull request by keith-turner**
Found this problem while testing #3143.  Since it was a more general problem, decided to fix it in 1.1

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/client/impl/TabletLocatorImpl.java [hunk 1]
```diff
@@ -330,8 +348,16 @@ else if (!lookupFailed)
         tabletLocations.add(tl);
       }
 
-      for (TabletLocation tl2 : tabletLocations) {
-        TabletLocatorImpl.addRange(binnedRanges, tl2.tablet_location, tl2.tablet_extent, range);
+      // Ensure the extents found are non overlapping and have no holes. When reading some extents
+      // from the cache and other from the metadata table in the loop above we may end up with
+      // non-contiguous extents. This can happen when a subset of exents are placed in the cache and
+      // then after that merges and splits happen.
+      if (isContiguous(tabletLocations)) {
```
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/client/impl/TabletLocatorImpl.java:355]: I'm wondering if this check should be moved to the end of loop `l1`. Multiple disjoint ranges could be sent into this method and I'm not sure that's taken into account here. I'm assuming that you want to check that you have a contiguous list of TabletLocations for each Range.

File: core/src/test/java/org/apache/accumulo/core/client/impl/TabletLocatorImplTest.java [hunk 2]
```diff
@@ -1115,6 +1116,82 @@ public void testBinRanges5() throws Exception {
...
+    ranges = nrl(nr("01", "17"));
+    runTest(tableName, ranges, metaCache, new HashMap<>(), nrl(nr("01", "17")));
+
+    // After the above test fails it should cause e3 to be added to the cache. Because e3 overlaps
+    // e2, when e3 is added then e2 is removed. Therefore, the following binRanges call should
+    // succeed and find the range overlaps e1 and e3.
+    expected = createExpectedBinnings("l1", nol(e1, nrl(nr("01", "17")), e3, nrl(nr("01", "17"))));
+    runTest(tableName, ranges, metaCache, expected, nrl());
+  }
+
```
Comment (dlmarion) [inline core/src/test/java/org/apache/accumulo/core/client/impl/TabletLocatorImplTest.java:1160]: I think we need a test that uses multiple non-contiguous ranges.

**Review by dlmarion — CHANGES_REQUESTED** [review]

File: core/src/test/java/org/apache/accumulo/core/client/impl/TabletLocatorImplTest.java (hunk 2 above)
Comment (keith-turner) [reply core/src/test/java/org/apache/accumulo/core/client/impl/TabletLocatorImplTest.java:1160]: I added a new test in 1b093f9

File: core/src/main/java/org/apache/accumulo/core/client/impl/TabletLocatorImpl.java (hunk 1 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/client/impl/TabletLocatorImpl.java:355]: The code is checking that the extents found for each range are contiguous.  If not, the range is added to the list of failed ranges.  I think the check is in the right place.  Looking back at the code I noticed it was not setting a boolean that other places set when adding a failed range, so I added that.

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/impl/TabletLocatorImpl.java:355]: Ok, you are right, I missed a bracket, so I thought that this was outside of loop `l1`.

**Review by dlmarion — APPROVED** [review]

**Pull request merged**